[JENKINS-29715] Control rescheduling with action

ikedam opened the pull request:
[JENKINS-29715](https://issues.jenkins-ci.org/browse/JENKINS-29715)
[Build-timeout plugin](https://wiki.jenkins-ci.org/display/JENKINS/Build-timeout+Plugin) plans a new feature to reschedule the build when timeout occurs.
(jenkinsci/build-timeout-plugin#44, jenkinsci/build-timeout-plugin#45).
The rescheduling feature is already provided by naginator plugin and we want to reuse that feature from build-timeout plugin.

This change introduces `NaginatorScheduleAction`.
You can make naginator reschedule a build by adding `NaginatorScheduleAction` to the build.

I changed to `NaginatorPublisher` to use that, and it allows naginator plugin works with [Flexible Publish plugin](https://wiki.jenkins-ci.org/display/JENKINS/Flexible+Publish+Plugin) ([JENKINS-23984](https://issues.jenkins-ci.org/browse/JENKINS-23984)).

Outline of the change:
* Introduced `NaginatorScheduleAction`, which provides following features:
    * Check the retrying count and decides whether to reschedule.
    * Always rerun all matrix children.
* `NaginatorListener` now decides whether to reschedule the build with `NaginatorScheduleAction`. `NaginatorListener` now does followings:
    * Calcuate retying counts.
    * Reschedule the build.
* `NaginatorPublisher` defines `NaginatorPublisher.Action` extending `NaginatorScheduleAction` which does almost all what was done in `NaginatorListener`. Many codes are copied from `NaginatorListener`:
    * Check the build result
    * Check the build log.

ikedam commented on src/main/java/com/chikli/hudson/plugin/naginator/NaginatorListener.java:
```diff
@@ -28,91 +20,52 @@
 
     @Override
     public void onCompleted(AbstractBuild<?, ?> build, TaskListener listener) {
```
many codes are moved to `NavigatorPublisher.Action`.

ikedam commented on src/main/java/com/chikli/hudson/plugin/naginator/NaginatorListener.java:
```diff
@@ -28,91 +20,52 @@
     }
 
+    @Deprecated
     public boolean canSchedule(Run build, NaginatorPublisher naginator) {
```
This method no longer used in naginator plugin.

KostyaSha replied:
could you add javadoc?

ikedam commented on src/main/java/com/chikli/hudson/plugin/naginator/NaginatorPublisher.java:
```diff
@@ -144,4 +164,102 @@ public Notifier newInstance(StaplerRequest req, JSONObject formData) throws Form
+        }
+        
+        @Override
+        public boolean shouldSchedule(Run<?, ?> run, TaskListener listener, int retryCount) {
```
almost all are copied from `NaginatorListener`.

ikedam commented on src/test/java/com/chikli/hudson/plugin/naginator/NaginatorPublisherTest.java:
```diff
@@ -0,0 +1,148 @@
+    }
+    
+    @Test
+    public void testMatrixBuild() throws Exception {
```
This is a regression test.
There was no test for matrix builds.

KostyaSha replied:
better add javadoc with description

KostyaSha commented on src/main/java/com/chikli/hudson/plugin/naginator/NaginatorListener.java:
```diff
@@ -28,91 +20,52 @@
+                        MatrixBuild mb = (MatrixBuild) build;
+                        List<MatrixRun> matrixRuns = mb.getRuns();
+    
+                        for(MatrixRun r : matrixRuns) {
```
`for (`

ikedam replied:
Oops. I added a commit to fix code formatting in 4e29dbb7f800c9399d12ce1f2f8cfcd2b0eb448a.

KostyaSha commented on src/main/java/com/chikli/hudson/plugin/naginator/NaginatorPublisher.java:
```diff
@@ -87,7 +98,16 @@ public int getMaxSchedule() {
     @Override
     public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
-        // Nothing to do during the build, see NaginatorListener
+        if (!(build instanceof MatrixRun)) {
```
reverse if to exlcude `!` ?

ikedam replied:
Added a commit cb42dcc2fa40bf3d0b80c8426ed7b1879ac57479 to reverse that.

KostyaSha commented on src/test/java/com/chikli/hudson/plugin/naginator/NaginatorPublisherTest.java:
```diff
@@ -0,0 +1,148 @@
+        }
+        
+        // only failed children are rescheduled if rerunMatrixPart is set.
+        {
```
Why not split into 2 tests?

ikedam replied:
* To save execution time of tests as Jenkins is launched for each tests.
* Those blocks both are to test the behavior for multi-configuration tests.

They are not so important and I can split it into 2 tests.
Do you like me splitting also tests in `NaginatorScheduleActionTest`?

KostyaSha replied:
> To save execution time of tests as Jenkins is launched for each tests.

Rule -> ClassRule

ikedam replied:
I didn't know `ClassRule`, thanks.
I'll split tests.

KostyaSha replied:
> Should I also split tests in NaginatorScheduleActionTest?

If their logic is not related, then imho better have logically splitted tests

lanwen replied:
please don't mix all test-cases in one test method. One method = one testcase. (Its test-code smell when more than one case in method) 

Please read the
```
XUnit Test Patterns
Refactoring Test Code
```

ikedam commented:
The test failure isn't caused by this change: https://jenkins.ci.cloudbees.com/job/plugins/job/naginator-plugin/org.jenkins-ci.plugins$naginator/60/testReport/junit/com.chikli.hudson.plugin.naginator/NaginatorListenerTest/testFailureWithRebuild/
* The assertion states the build was not rescheduled.
```
junit.framework.AssertionFailedError: expected:<true> but was:<false>
	at junit.framework.Assert.fail(Assert.java:57)
	at junit.framework.Assert.failNotEquals(Assert.java:329)
	at junit.framework.Assert.assertEquals(Assert.java:78)
	at junit.framework.Assert.assertEquals(Assert.java:174)
	at junit.framework.Assert.assertEquals(Assert.java:181)
	at junit.framework.TestCase.assertEquals(TestCase.java:341)
	at com.chikli.hudson.plugin.naginator.NaginatorListenerTest.testFailureWithRebuild(NaginatorListenerTest.java:81)
```
* But the build is rescheduled and run twice from the log:
```
Aug 01, 2015 11:15:17 PM hudson.model.Run execute
INFO: test0 #1 main build action completed: FAILURE
Aug 01, 2015 11:15:18 PM hudson.model.Run execute
INFO: test0 #2 main build action completed: FAILURE
```
* That might be caused for the 2nd run takes more than 1 second to finish. Codes in `NaginatorListenerTest`:
```
    private boolean isScheduledForRetry(FreeStyleProject project) throws InterruptedException, ExecutionException {
        FreeStyleBuild build = project.scheduleBuild2(0).get();
        Thread.sleep(1000); // wait for job to run and possibly re-run

        return project.getLastBuild().getNumber() > 1;
    }
```

KostyaSha commented on src/main/java/com/chikli/hudson/plugin/naginator/NaginatorPublisher.java:
```diff
@@ -144,4 +164,102 @@ public Notifier newInstance(StaplerRequest req, JSONObject formData) throws Form
+        }
+
+        @Override
+        public boolean shouldScheduleForMatrixRun(MatrixRun run, TaskListener listener) {
```
annotate `run` with `@Nonnull` and everywhere else

ikedam replied:
Done with 55c8d827099553a7d03c40662658c62f36117798.
I'm stranger in `@Nonnull` and `@Nullable` annotations and would you review that?

lanwen commented on src/test/java/com/chikli/hudson/plugin/naginator/NaginatorScheduleActionTest.java:
```diff
@@ -0,0 +1,335 @@
+    
+    @Test
+    public void testMatrix() throws Exception {
+        // all children are rescheduled.
```
please don't mix all test-cases in one test method. One method = one testcase.

KostyaSha commented on src/test/java/com/chikli/hudson/plugin/naginator/NaginatorScheduleActionTest.java:
```diff
@@ -0,0 +1,335 @@
+        FreeStyleProject p = j.createFreeStyleProject();
+        FreeStyleBuild b = p.scheduleBuild2(0).get();
+        
+        {
```
Why do you need anonymous block code?
You can just separate lines into groups of lines with new line separator or add some comment before group of lines

KostyaSha commented:
> The rescheduling feature is already provided by naginator plugin and we want to reuse that feature from build-timeout plugin.

infra is down, can't check documentation. Why there are 2 plugins? Why not merge functionality into one?

lanwen commented on src/test/java/com/chikli/hudson/plugin/naginator/NaginatorScheduleActionTest.java:
```diff
@@ -0,0 +1,335 @@
+        
+        {
+            NaginatorScheduleAction target = new NaginatorScheduleAction(2);
+            assertTrue(target.shouldSchedule(b, TaskListener.NULL, 0));
```
please add comments to this asserts. Or we can't know which line will fail

ikedam replied:
Let me see.
Do comments mean code comments or messages for assertions (the first `String` parameters of `assertXxx`)?
I splitted tests and 1 test method contains 1 test case. I think that clarified what each test method asserts. Did that resolve the priblem?

ikedam commented:
I split tests so that 1test method = 1 test cases. This also removed all anonymous blocks.
I believe this resolved all existing comments for test codes.
Please let me know if I missed any of them.

ikedam commented:
> infra is down, can't check documentation. Why there are 2 plugins? Why not merge functionality into one?

Build-timeout plugin: Watches timeouts of builds. It triggers timeout operations when timeout occurs. The default operation is to abort the build.
Naginator-plugin: Reschedule a build (when it fails).

Those basic functionality is differ and it makes sense that there are two plugins.

Build-timeout plugin can have a build failed when timeout occurs so that the build is rescheduled by naginator plugin.
However it's more useful if one can reschedule a build only when timeout occurs.
Users  may not want to reschedule builds that apparently failed.

ikedam commented on src/main/java/com/chikli/hudson/plugin/naginator/NaginatorPublisher.java:
```diff
@@ -144,4 +167,99 @@ public Notifier newInstance(StaplerRequest req, JSONObject formData) throws Form
+            return true;
+        }
+        
+        private boolean parseLog(File logFile, @Nonnull String regexp) throws IOException {
```
`Run#getLogFile` is annotated with `@Nonnull` since Jenkins 1.568.
Naginator plugin targets Jenkins 1.554 and `logFile` is not `@Nonnull` for now.
Should I add `@Nullable` and null-checking codes in this case?

KostyaSha replied:
Not really critical imho, but you can mark //TODO since 1.554 baseline remove annotation

KostyaSha replied:
And this annotations will be used only on pom jenkins version, because they are required only for build time. So you should add them and remove after base upgrade.

KostyaSha commented on src/test/java/com/chikli/hudson/plugin/naginator/NaginatorScheduleActionTest.java:
```diff
@@ -0,0 +1,408 @@
+     * 
+     * @throws Exception
+     */
+    public void testMatrixWithoutRerunMatrixPart() throws Exception {
```
did you forget `@Test` for methods?

ikedam replied:
Oops!
I often do that mistake...I'll immediately fix that.

KostyaSha commented on src/test/java/com/chikli/hudson/plugin/naginator/NaginatorScheduleActionTest.java:
```diff
@@ -0,0 +1,408 @@
+     * 
+     * @throws Exception
+     */
+    public void testMatrixWithNoChildren() throws Exception {
```
`@Test`?

KostyaSha commented on src/test/java/com/chikli/hudson/plugin/naginator/NaginatorScheduleActionTest.java:
```diff
@@ -0,0 +1,408 @@
+     * 
+     * @throws Exception
+     */
+    public void testMatrixWithRerunMatrixPartWithFiltering() throws Exception {
```
`@Test`?

KostyaSha commented on src/main/java/com/chikli/hudson/plugin/naginator/NaginatorScheduleAction.java:
```diff
@@ -0,0 +1,125 @@
+     */
+    public NaginatorScheduleAction(int maxSchedule, @Nullable ScheduleDelay delay, boolean rerunMatrixPart) {
+        this.maxSchedule = maxSchedule;
+        this.delay = (delay != null)?delay:new ProgressiveDelay(5 * 60, 3 * 60 * 60);
```
please put spaces before and after `? : `

KostyaSha commented on src/main/java/com/chikli/hudson/plugin/naginator/NaginatorListener.java:
```diff
@@ -27,92 +21,53 @@
+    public void onCompleted(AbstractBuild<?, ?> build, @Nonnull TaskListener listener) {
         // Do nothing for a single Matrix run. (Run only when all Matrix finishes)
-        if (build instanceof MatrixRun) {
+        if (build == null || build instanceof MatrixRun) {
```
AFAIK instanceof is already doing NPE, so this change should be reverted

ikedam replied:
`null instanceof SomeClass` always returns `false`: http://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.20.2
(though, that sounds a little strange to me as `MatrixRun` can be `null`...)
This null-check is not for `instanceof` but for following codes (e.g. `calculateRetryCount`).

The code comment above may be confusing and I'll update that.

KostyaSha replied:
sorry, read wrong, thought it `!=` because initial code was reversed

ikedam commented:
Added new commits to fix problems pointed in the review: fd6e1cdfa73157bf521d4eef9cddd0a8630337e2, 15ea3fe2b1e8d8b97e4671436878e2cff4179ff0

KostyaSha commented on src/main/java/com/chikli/hudson/plugin/naginator/NaginatorPublisher.java:
```diff
@@ -144,4 +167,101 @@ public Notifier newInstance(StaplerRequest req, JSONObject formData) throws Form
 
     private static final Logger LOGGER = Logger.getLogger(NaginatorPublisher.class.getName());
 
+    public static class Action extends NaginatorScheduleAction {
```
Please call it `NaginatorAction`? I was really confused that `Action` comes from core

KostyaSha replied:
Is it possible to place it in separate class?

KostyaSha commented on src/main/java/com/chikli/hudson/plugin/naginator/NaginatorPublisher.java:
```diff
@@ -144,4 +167,101 @@ public Notifier newInstance(StaplerRequest req, JSONObject formData) throws Form
+            this.checkRegexp = publisher.isCheckRegexp();
+        }
+        
+        @Nullable
```
as far as i understand FB, any field is Nullabble by default, if you want enforce null check, then place `CheckForNull`

KostyaSha commented:
@ikedam idea to use Actions looks good for me, could you answer on new comments and squash commits into one?

ikedam commented:
Squashed commits into 586b71a3002f3459a39c92c0be11f213447d2af8 .

You can review the branch before squashing with 
https://github.com/ikedam/naginator-plugin/tree/feature/JENKINS-29715_ControlWithActionBeforeSquash

New added commits are:
95727dc6bc955e8e8429df81970fd4ef96c5d936
ce9c88c840e5c64c6ecc4c1ef0e827bedd27a298

KostyaSha commented on src/main/java/com/chikli/hudson/plugin/naginator/NaginatorPublisherScheduleAction.java:
```diff
@@ -0,0 +1,121 @@
+            }
+        }
+    }
+}
```
add new line :(

ikedam commented:
Oops, sorry.
Added a new commit to added a line break (372e55ef39a505d184debcd3fab24a91f52a5332) and squashed again (9621e5d8f41158771440eab2677796cb331f4f03).

KostyaSha commented:
LGTM, do you want to do release?

ikedam commented:
Would you make a snapshot release to let build-timeout-plugin start to develop the integration with naginator-plugin?

An actual new release (1.16?) can be done after resolving other pull requests.
And I also plan to work for [JENKINS-17626](https://issues.jenkins-ci.org/browse/JENKINS-17626).

KostyaSha commented:
@ikedam there is alpha/beta releases. Released 1.16-alpha-1 that you can use in reference implementations.

PS will you be able to resolve neighbour PRs related to matrix builds?

ikedam commented:
Thanks for the release.

> PS will you be able to resolve neighbour PRs related to matrix builds?

No, I won't do that.
